File: rust/testsuite/src/services_impl/datastore/datastore_container_initializer.rs
use kurtosis_rust_lib::services::{docker_container_initializer, service_context::{ServiceContext}};
use std::{collections::{HashSet, HashMap}, path::PathBuf};
use crate::services_impl::datastore::datastore_service::DatastoreService;
use std::fs::File;
use anyhow::Result;

const PORT: u32 = 1323;
const PROTOCOL: &str = "tcp";
const TEST_VOLUME_MOUNTPOINT: &str = "/test-volume";

pub struct DatastoreContainerInitializer {
    docker_image: String,
}

impl DatastoreContainerInitializer {
    pub fn new(docker_image: &str) -> DatastoreContainerInitializer {
        return DatastoreContainerInitializer{
            docker_image: docker_image.to_owned(),
        };
    }
}

impl docker_container_initializer::DockerContainerInitializer<DatastoreService> for DatastoreContainerInitializer {
    fn get_docker_image(&self) -> &str {
        return &self.docker_image;
    }

    fn get_used_ports(&self) -> HashSet<String> {
        let mut result = HashSet::new();
        result.insert(format!("{}/{}", PORT, PROTOCOL));
        return result;
    }

    fn get_service(&self, service_context: ServiceContext) -> Box<dyn kurtosis_rust_lib::services::service::Service> {
        let service = DatastoreService::new(
            service_context,
            PORT);
        return Box::new(service);
    }

    fn get_files_to_generate(&self) -> HashSet<String> {
        return HashSet::new();
    }

    fn initialize_generated_files(&self, _: HashMap<String, File>) -> Result<()> {
        return Ok(());
    }

    fn get_files_artifact_mountpoints(&self) -> HashMap<String, String> {
        return HashMap::new();
    }


    fn get_test_volume_mountpoint(&self) -> &'static str {
        return TEST_VOLUME_MOUNTPOINT;
    }

    fn get_start_command_overrides(
            &self,
            _: HashMap<String, PathBuf>, 
            _: &str
    ) -> Result<(Option<Vec<String>>, Option<Vec<String>>)> {
        // We have a launch CMD specified in the Dockerfile the datastore service was built with and we don't need
        // to specify an ENTRYPOINT, so we leave everything nil
        return Ok((None, None))
    }

}

File: golang/testsuite/services_impl/datastore/datastore_container_initializer.go
/*
 * Copyright (c) 2020 - present Kurtosis Technologies LLC.
 * All Rights Reserved.
 */

package datastore

import (
	"fmt"
	"github.com/kurtosis-tech/kurtosis-libs/golang/lib/services"
	"os"
)

const (
	port = 1323

	testVolumeMountpoint = "/test-volume"
)

type DatastoreContainerInitializer struct {
	dockerImage string
}

func NewDatastoreContainerInitializer(dockerImage string) *DatastoreContainerInitializer {
	return &DatastoreContainerInitializer{dockerImage: dockerImage}
}

func (d DatastoreContainerInitializer) GetDockerImage() string {
	return d.dockerImage
}

func (d DatastoreContainerInitializer) GetUsedPorts() map[string]bool {
	return map[string]bool{
		fmt.Sprintf("%v/tcp", port): true,
	}
}

func (d DatastoreContainerInitializer) GetServiceWrappingFunc() func (ctx *services.ServiceContext) services.Service {
	return func(ctx *services.ServiceContext) services.Service {
		return NewDatastoreService(ctx, port);
	}
}

func (d DatastoreContainerInitializer) GetFilesToGenerate() map[string]bool {
	return map[string]bool{}
}

func (d DatastoreContainerInitializer) InitializeGeneratedFiles(filesToGenerate map[string]*os.File) error {
	return nil
}

func (d DatastoreContainerInitializer) GetFilesArtifactMountpoints() map[services.FilesArtifactID]string {
	return map[services.FilesArtifactID]string{}
}

func (d DatastoreContainerInitializer) GetTestVolumeMountpoint() string {
	return testVolumeMountpoint
}

func (d DatastoreContainerInitializer) GetStartCommandOverrides(
		mountedFileFilepaths map[string]string,
		ipAddr string) (entrypointArgs []string, cmdArgs []string, resultErr error) {
	// We have a launch CMD specified in the Dockerfile the datastore service was built with and we don't need
	// to specify an ENTRYPOINT, so we leave everything nil
	return nil, nil, nil
}

